RFC-0001-torch-function-for-methods.md

```diff
@@ -0,0 +1,343 @@
+
+1. Support subclassing `torch.Tensor` in Python
+2. Preserve `torch.Tensor` subclasses when calling `torch` functions on them
+3. Use the PyTorch API with `torch.Tensor`-like objects that are _not_ `torch.Tensor`
+   subclasses
+4. Preserve `torch.Tensor` subclasses when calling `torch.Tensor` methods.
+5. Propagating subclass instances correctly also with operators, using
+   views/slices/indexing/etc.
+6. Preserve subclass attributes when using methods or views/slices/indexing.
+7. A way to insert code that operates on both functions and methods uniformly
+   (so we can write a single function that overrides all operators).
+8. The ability to give external libraries a way to also define
+   functions/methods that follow the `__torch_function__` protocol.
+
+Goals 1‒6 are explicitly about subclassing, goal 7 is already partially achieved via the `__torch_function__` protocol (which we're proposing to extend to methods), and goal 8 is a by-product required to make overridden `torch.Tensor` subclass methods behave similar to `torch.Tensor` methods.
+
+Achieving interoperability with NumPy and adopting its array protocols is out
+of scope for this proposal and we propose to defer it to a later proposal.
+
+We propose to solve this problem with the following changes to PyTorch:
+
+1. Make methods and operators of `torch.Tensor` go through the
+   `__torch_function__` machinery.
+2. Add a `types` argument to `__torch_function__`, to make it match NumPy's
+   `__array_function__`.
+3. Make `torch.Tensor._make_subclass` public API by renaming it to `torch.Tensor.make_subclass`.
+4. Make `torch.Tensor` gain a generic implementation of `__torch_function__`.
+
+## Usage and Impact
+Once this proposal is merged, users of subclasses of `torch.Tensor` will have
+a much more streamlined experience. Namely, the following code example will
+work as-is, without the need for any further modification:
+
+```python
+class SubTensor(torch.Tensor):
+    a = 1
+
+t = SubTensor([1])
+s = t.sum()
+isinstance(s, SubTensor)  # True
+s.a  # 1
+i = t[0]
+isinstance(i, SubTensor)  # True
+i.a  # 1
+
+s2 = t + torch.Tensor(1)
+isinstance(s2, SubTensor)  # True
+s2.a  # 1
+
+s3 = torch.Tensor(1) + t
+isinstance(s3, SubTensor)  # True
+s3.a  # 1
+```
+
+Additionally, it will provide subclass authors the ability to also modify the
+results of methods and operators in `__torch_function__`, along with regular
+function calls, and to modify the result to their specific use-case, perform
+logging, or otherwise change the result or the action of the method. For
+example:
+
+```python
+import logging
+
+class LoggingTensor(torch.Tensor):
+    def __torch_function__(self, func, types, args, kwargs):
+        logging.info(f"func: {func.__name__}, args: {args!r}, kwargs: {kwargs!r}")
+        return super().__torch_function__(
+            func,
+            tuple(
+                t for t in types if not issubclass(t, LoggingTensor)
```

Comment:
It seems annoying that all users have to write this boilerplate whenever they make parent calls.

Reply:
BTW, I don't think we need to design for this here, but I am wondering: do you know if this pattern works correctly for multiple inheritance in Python?

Reply:
I think we could provide an `exclude_type` helper function, but beyond that, I'm not sure what we can do.

Reply:
And it will, you can just add `type(self).__bases__[base_index_you_want_here]` to the list of `types`, and assuming it implements `__torch_function__` and is hit in the MRO, it'll work fine.